Push-style event: Delete subscriber after retry failures

sunharis opened the pull request:
Push style eventing implements sendEvent retries based on the
maximum retry limits set at the event service. When the max retry
is hit, it means that the subscriber is gone and bmc need not
attempt sending the event to that suscriber when next event occurs.
BMC should delete the subcription in this case.

Current implementation does not delete the subscription on hitting
maximum number of sendEvent retries. This overloads the http_client
at bmc to attempt a sendEvent to a non-existing or stale subscription.

This commit implements
1. Delete subscription after max retires
2. Store subscription id as a member of subscriber
3. Some trace enhancements at http client

Tested by:
 1. Subscribe to events and close the subscriber abruptly to see
    bmc retries and after all retry attempt failures, stops sending
    the events to that subscriber
 2. A bmc restart should cleanup these stale subscribers
 3. Tested sanity of the existing push-style eventing feature

Signed-off-by: Sunitha Harish <sunharis@in.ibm.com>
Change-Id: Idae9b4b4a8a4cc25239ea4637b7bde9f2e380b4f

sunharis commented:
The upstream bmcweb does not have the https client enabled with ssl. So the code is considerably different. I will create an upstream commit separately with this change. Same code can not be pushed upstream

raviteja-b commented on include/event_service_store.hpp:
```diff
@@ -258,6 +258,32 @@ class EventServiceStore
+    }
+    void removeSubscription(const std::string& id)
+    {
+        BMCWEB_LOG_ERROR << "removeSubscription: " << id;
```
Info log?

sunharis replied:
removed this

gtmills commented on redfish-core/include/event_service_manager.hpp:
```diff
@@ -559,8 +563,15 @@ class Subscription : public persistent_data::UserSubscription
+    void setSubId(const std::string& id)
+    {
+        subId = id;
+        BMCWEB_LOG_ERROR << "subId saved as: " << subId;
```
Why Error here?

sunharis replied:
removed this

gtmills reviewed (commented):
Seems like we added a lot of BMCWEB_LOG_ERROR, do we really need these all?

sunharis commented:
> Seems like we added a lot of BMCWEB_LOG_ERROR, do we really need these all?

I added some at the retry path and while removing the subscription, so that they are logged at the journal traces

gtmills commented on include/event_service_store.hpp:
```diff
@@ -258,6 +258,31 @@ class EventServiceStore
+            needWrite = true;
+            return;
+        }
+        BMCWEB_LOG_ERROR << "Subscriber " << id
```
Is this really needed as an LOG ERROR?

sunharis replied:
This is hit if there is an error actually. ie the subscriber is asked to be removed, but its not found at the subscription config map

gtmills commented on include/event_service_store.hpp:
```diff
@@ -258,6 +258,31 @@ class EventServiceStore
+        auto obj = subscriptionsConfigMap.find(id);
+        if (obj != subscriptionsConfigMap.end())
+        {
+            BMCWEB_LOG_ERROR << "Deleting " << id
```
Is this really needed as an LOG ERROR since you have 

                BMCWEB_LOG_ERROR << "TerminateAfterRetries is set. retryCount: "
                                 << retryCount << " .Subscriber: " << subId
                                 << " deleted";
                                 
                                 above?

gtmills commented on http/http_client.hpp:
```diff
@@ -366,12 +367,20 @@ class HttpClient : public std::enable_shared_from_this<HttpClient>
-                // TODO: delete subscription
                 state = ConnState::terminated;
+                // Remove the subscription
+                BMCWEB_LOG_ERROR << "TerminateAfterRetries is set. retryCount: "
```
Can we have this one or the one in event_service_store.hpp on 277?
